OpenRA.Game/FileSystem/InstallShieldCABExtractor.cs

```diff
@@ -0,0 +1,294 @@
+		{
+			s = GlobalFileSystem.Open(filename);
+			var buff = new List<char>(filename.Substring(0, filename.LastIndexOf('.')).ToCharArray());
+			for (int i = buff.Count - 1; char.IsNumber(buff[i]); --i) {
+				buff.RemoveAt(i);
+			}
+
+			commonPath = new string(buff.ToArray());
+			var reader = new BinaryReader(s);
+			var signature = reader.ReadUInt32();
+			if (signature != 0x28635349) throw new InvalidDataException("Not an Installshield CAB package");
+			commonHeader = new CommonHeader(reader);
+			cabDescriptor = new CabDescriptor(reader, commonHeader);
+			reader.BaseStream.Seek(commonHeader.CabDescriptorOffset + cabDescriptor.FileTableOffset, SeekOrigin.Begin);
+
+			directoryTable = new List<uint>();
+			for (uint i = cabDescriptor.DirectoryCount; i > 0; --i) {
+				directoryTable.Add(reader.ReadUInt32());
+			}
+
+			fileTable = new List<uint>();
+			for (uint i = cabDescriptor.FileCount; i > 0; --i) {
+				fileTable.Add(reader.ReadUInt32());
+			}
+
+			directoryNames  = new Dictionary<uint, string>();
+			fileDescriptors = new Dictionary<uint, FileDescriptor>();
+		}
+
+		public string DirectoryName(uint index)
+		{
+			if (directoryNames.ContainsKey(index))
+				return directoryNames[index];
+			var reader = new BinaryReader(s);
+			reader.BaseStream.Seek(commonHeader.CabDescriptorOffset +
+					cabDescriptor.FileTableOffset +
+					directoryTable[(int)index],
+					SeekOrigin.Begin);
+			var sb = new System.Text.StringBuilder();
+			byte c = reader.ReadByte();
+			while (c != 0) {
+				sb.Append((char)c);
+				c = reader.ReadByte();
+			}
+
+			return sb.ToString();
+		}
+
+		public uint DirectoryCount() {
+			return cabDescriptor.DirectoryCount;
+		}
+
+		public string FileName(uint index)
+		{
+			if (!fileDescriptors.ContainsKey(index))
+				AddFileDescriptorToList(index);
+			return fileDescriptors[index].Filename;
+		}
+
+		void AddFileDescriptorToList(uint index) {
+			var reader = new BinaryReader(s);
+			reader.BaseStream.Seek(commonHeader.CabDescriptorOffset +
+					cabDescriptor.FileTableOffset +
+					cabDescriptor.FileTableOffset2 +
+					index * 0x57,
+					SeekOrigin.Begin);
+			var fd = new FileDescriptor(reader,
+				commonHeader.CabDescriptorOffset + cabDescriptor.FileTableOffset);
+			fileDescriptors.Add(index, fd);
+		}
+
+		public uint FileCount()
+		{
+			return cabDescriptor.FileCount;
+		}
+
+		public void ExtractFile(uint index, string fileName)
+		{
+			if (!fileDescriptors.ContainsKey(index))
+				AddFileDescriptorToList(index);
+			var fd = fileDescriptors[index];
+			if ((fd.Flags & FILEINVALID) != 0) throw new Exception("File Invalid");
+			if ((fd.LinkFlags & LINKPREV) != 0)
+			{
+				ExtractFile(fd.LinkPrevious, fileName);
+				return;
+			}
+
+			if ((fd.Flags & FILESPLIT) != 0 || (fd.Flags & FILEOBFUSCATED) != 0)
+				throw new Exception("Haven't implemented");
+
+			var fil = GlobalFileSystem.Open(string.Format("{0}{1}.cab", commonPath, fd.Volume));
```

Comment:
You forgot to dispose this stream so you are leaking resources. In C# you can use
```csharp
using (var fil = GlobalFileSystem.Open(string.Format("{0}{1}.cab", commonPath, fd.Volume)))
	using (var reader = new BinaryReader(fil))
	{
		reader.ReadUInt32() [...]
	}
```

to automatically clean up after usage.